runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.spark.InterruptibleIterator;
+import org.apache.spark.Partition;
+import org.apache.spark.SparkContext;
+import org.apache.spark.TaskContext;
+import org.apache.spark.api.java.JavaSparkContext$;
+import org.apache.spark.rdd.RDD;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Classes implementing Beam {@link Source} {@link RDD}s.
+ */
+public class SourceRDD {
+
+  /**
+   * A SourceRDD.Bounded reads input from a {@link BoundedSource} and creates a Spark {@link RDD}.
+   * This is the default way for the SparkRunner to read data from Beam's BoundedSources.
+   */
+  public static class Bounded<T> extends RDD<WindowedValue<T>> {
+    private static final Logger LOG = LoggerFactory.getLogger(SourceRDD.Bounded.class);
+
+    private final BoundedSource<T> source;
+    private final SparkRuntimeContext runtimeContext;
+    private final int numPartitions;
+
+    // to satisfy Scala API.
+    private static final scala.collection.immutable.List<Dependency<?>> NIL =
+        scala.collection.immutable.List.empty();
+
+    public Bounded(SparkContext sc,
+                   BoundedSource<T> source,
+                   SparkRuntimeContext runtimeContext) {
+      super(sc, NIL, JavaSparkContext$.MODULE$.<WindowedValue<T>>fakeClassTag());
+      this.source = source;
+      this.runtimeContext = runtimeContext;
+      // the input parallelism is determined by Spark's scheduler backend.
+      // when running on YARN/SparkDeploy it's the result of max(totalCores, 2).
+      // when running on Mesos it's 8.
+      // when running local it's the total number of cores (local = 1, local[N] = N,
+      // local[*] = estimation of the machine's cores).
+      // ** the configuration "spark.default.parallelism" takes precedence over all of the above **
+      this.numPartitions = sc.defaultParallelism();
+      checkArgument(this.numPartitions > 0, "Number of partitions must be greater than zero.");
+    }
+
+    private static final long DEFAULT_BUNDLE_SIZE = 64 * 1024 * 1024;
+
+    @Override
+    public Partition[] getPartitions() {
+      long desiredSizeBytes = DEFAULT_BUNDLE_SIZE;
+      try {
+        desiredSizeBytes = source.getEstimatedSizeBytes(
+            runtimeContext.getPipelineOptions()) / numPartitions;
+      } catch (Exception e) {
+        LOG.warn("Failed to get estimated size of bundle, default is {} bytes.",
+            DEFAULT_BUNDLE_SIZE);
+      }
+      try {
+        List<? extends Source<T>> partitionedSources = source.splitIntoBundles(desiredSizeBytes,
+            runtimeContext.getPipelineOptions());
+        Partition[] partitions = new SourcePartition[partitionedSources.size()];
+        for (int i = 0; i < partitionedSources.size(); i++) {
+          partitions[i] = new SourcePartition<>(id(), i, partitionedSources.get(i));
+        }
+        return partitions;
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to create partitions.", e);
```

Comment:
"...for Source %s" (maybe "for Source of type " + source.getClass().getSimpleName())